models/contracts/optimism/contracts_optimism_contract_mapping.sql

```diff
@@ -0,0 +1,266 @@
+    ,t.symbol
+  from base_level as bl 
+  join {{ ref('tokens_optimism_erc20') }} as t
+    on bl.contract_address = t.contract_address
+  group by 1, 2
+
+  union all 
+
+  select 
+    bl.contract_address
+    ,t.name as symbol
+  from base_level as bl 
+  join {{ ref('tokens_optimism_nft') }} as t
+    on bl.contract_address = t.contract_address
+  group by 1, 2
+)
+-- starting from 0 
+{% for i in range(max_levels) -%}
+,level{{i}} as (
+    select
+      {{i}} as level 
+      ,coalesce(u.creator_address, b.creator_address) as creator_address
+      {% if loop.first -%}
+      ,case
+        when u.creator_address is NULL then NULL
+        else b.creator_address
+      end as contract_factory
+      {% else -%}
+      ,case
+        when u.creator_address is NULL then b.contract_factory
+        else b.creator_address
+      end as contract_factory
+      {% endif %}
+      ,b.contract_address
+      ,b.created_time
+      ,b.creation_tx_hash
+      ,b.is_self_destruct
+    {% if loop.first -%}
+    from base_level as b
+    left join base_level as u
+      on b.creator_address = u.contract_address
+    {% else -%}
+    from level{{i-1}} as b
+    left join base_level as u
+      on b.creator_address = u.contract_address
+    {% endif %}
+)
+{%- endfor %}
+
+,creator_contracts as (
+  select 
+    f.creator_address
+    ,f.contract_factory
+    ,f.contract_address
+    ,coalesce(cc.contract_project, ccf.contract_project) as contract_project 
+    ,f.created_time
+    ,f.is_self_destruct
+    ,f.creation_tx_hash
+  from level{{max_levels - 1}} as f
+  left join {{ ref('contracts_optimism_contract_creator_address_list') }} as cc 
+    on f.creator_address = cc.creator_address
+  left join {{ ref('contracts_optimism_contract_creator_address_list') }} as ccf
+    on f.contract_factory = ccf.creator_address
+  where f.contract_address is not null
+ )
+,combine as (
+  select 
+    cc.creator_address
+    ,cc.contract_factory
+    ,cc.contract_address
+    ,coalesce(cc.contract_project, oc.namespace) as contract_project 
+    ,oc.name as contract_name 
+    ,cc.created_time
+    ,coalesce(cc.is_self_destruct, false) as is_self_destruct
+    ,'creator contracts' as source
+    ,cc.creation_tx_hash
+  from creator_contracts as cc 
+  left join {{ source('optimism', 'contracts') }} as oc 
+    on cc.contract_address = oc.address 
+
+  union all
+  -- ovm 1.0 contracts
+
+  select 
+    creator_address
+    ,NULL as contract_factory
+    ,contract_address
+    ,contract_project
+    ,contract_name
+    ,to_timestamp(created_time) as created_time
+    ,false as is_self_destruct
+    ,'ovm1 contracts' as source
+    ,NULL as creation_tx_hash
+  from {{ source('ovm1_optimism', 'contracts') }} as c
+  where 
+    true
+    {% if is_incremental() %} -- this filter will only be applied on an incremental run 
+    and not exists (
+      select 1
+      from {{ this }} as gc
+      where 
+        gc.contract_address = c.contract_address
+        and (
```

Comment:
177-178 seems duplicative now. Do you see a case for having this? May be a hold over from a prior version, this should just be checking for "does this contract already exist."

Reply:
ignore me, can't read